File: python/app/vulns/deserialization.py
"""Insecure Deserialization / Software & Data Integrity Failures.

OWASP 2021 A08 Software and Data Integrity Failures  ->  OWASP 2025 A08.

This module mirrors the comment/annotation format of the reference SQL-injection
module: each permutation carries a standard header describing the flaw, its CWE,
an exploitation-likelihood rating, prevalence today, an example payload, and the
corresponding safe pattern. Every dangerous sink is marked ``VULNERABLE:`` and
the file ends with ONE clearly-labelled SAFE reference handler.

All handlers accept the named field from a JSON body, a form field, or a query
arg, so any of these work:

    curl -X POST /deserialization/pickle -d 'data=<base64>'
    curl -X POST /deserialization/pickle -H 'Content-Type: application/json' -d '{"data":"<base64>"}'
"""
import base64
import json
import pickle

import yaml
from flask import Blueprint, jsonify, request

bp = Blueprint("deserialization", __name__, url_prefix="/deserialization")


def _field(name: str, default: str = "") -> str:
    """Pull a field from a JSON body, form body, or query string (in that order)."""
    body = request.get_json(silent=True) or {}
    if name in body:
        return body[name]
    return request.form.get(name, request.args.get(name, default))


# =============================================================================
# PERMUTATION 1 — Native pickle deserialization of attacker-controlled bytes
# OWASP 2021 A08 Software and Data Integrity Failures  ->  OWASP 2025 A08 Software and Data Integrity Failures
# CWE-502: Deserialization of Untrusted Data
# EXPLOITATION LIKELIHOOD: HIGH — pre-auth, deterministic RCE. pickle.loads runs
#   the object's __reduce__ during unpickling, so a self-contained payload (no
#   external gadget chain needed) executes arbitrary code; only caveat is the
#   attacker must craft/encode the pickle, which is a trivial 5-line script.
# PREVALENCE TODAY: greenfield MEDIUM (web frameworks avoid pickle, but the
#   Python ML/data-science stack still pickles models, joblib artifacts, Celery/
#   Redis task payloads and caches by default) | legacy/10yr tech-debt HIGH
#   (pickle-backed sessions, caches and RPC are everywhere in old Python code).
# EXAMPLE — build a payload whose __reduce__ shells out, then POST it:
#   python -c "import pickle,base64,subprocess as s;\
#   class E:\
#     def __reduce__(self): return (s.check_output, (['id'],))\
#   print(base64.b64encode(pickle.dumps(E())).decode())"
#   curl -X POST localhost:5000/deserialization/pickle -d "data=<paste base64>"
#   -> {"result": "b'uid=0(root) ...\\n'", ...}
# FIX: never unpickle untrusted input; use a data-only format (json.loads) — see
#   the SAFE reference handler below.
# =============================================================================
@bp.post("/pickle")
def pickle_load():
    data = _field("data")
    raw = base64.b64decode(data)
    # VULNERABLE: pickle.loads executes __reduce__ on attacker-controlled bytes -> RCE.
    obj = pickle.loads(raw)
    return jsonify({"deserialized_type": type(obj).__name__, "result": repr(obj)})


# =============================================================================
# PERMUTATION 2 — Unsafe config-format deserialization (yaml.unsafe_load)
# OWASP 2021 A08 Software and Data Integrity Failures  ->  OWASP 2025 A08 Software and Data Integrity Failures
# CWE-502: Deserialization of Untrusted Data
# EXPLOITATION LIKELIHOOD: HIGH — deterministic RCE with a one-line payload.
#   yaml.unsafe_load honours the !!python/object/apply tag, which constructs
#   arbitrary objects and CALLS arbitrary functions during parsing; no gadget
#   hunting required because the YAML names the callable directly.
# PREVALENCE TODAY: greenfield LOW (PyYAML has defaulted to safe loading since
#   5.1 and yaml.load without a Loader now warns; you must explicitly reach for
#   unsafe_load/Loader=UnsafeLoader) | legacy/10yr tech-debt MEDIUM-HIGH (heaps
#   of pre-5.1 code call bare yaml.load(), which used to be unsafe by default).
# EXAMPLE:
#   curl -X POST localhost:5000/deserialization/yaml \
#     --data-urlencode 'data=!!python/object/apply:subprocess.check_output [["id"]]'
#   -> {"result": "b'uid=0(root) ...\\n'", ...}
# FIX: yaml.safe_load(...) (SafeLoader) — see the SAFE reference handler below.
# =============================================================================
@bp.post("/yaml")
def yaml_load():
    data = _field("data")
    # VULNERABLE: unsafe_load constructs/invokes arbitrary Python objects via YAML tags -> RCE.
    obj = yaml.unsafe_load(data)
    return jsonify({"deserialized_type": type(obj).__name__, "result": repr(obj)})


# =============================================================================
# PERMUTATION 3 — Integrity failure via eval() of a user "formula"/config value
# OWASP 2021 A08 Software and Data Integrity Failures  ->  OWASP 2025 A08 Software and Data Integrity Failures
# CWE-95: Improper Neutralization of Directives in Dynamically Evaluated Code ('Eval Injection')
# EXPLOITATION LIKELIHOOD: CRITICAL — eval() runs the string as Python, so a
#   plain expression like __import__('os').popen('id').read() yields immediate,
#   reliable RCE with zero tooling and no gadget; trivially discovered and abused.
# PREVALENCE TODAY: greenfield LOW (eval on request data is a textbook red flag
#   that linters/SAST/code-review catch, and safe expression libs like ast or
#   simpleeval exist) | legacy/10yr tech-debt MEDIUM (hand-rolled calculators,
#   rules engines and "dynamic config" that eval() a stored formula still lurk).
# EXAMPLE:
#   curl -X POST localhost:5000/deserialization/eval \
#     --data-urlencode "expr=__import__('os').popen('id').read()"
#   -> {"result": "'uid=0(root) ...\\n'", "expr": "__import__('os')..."}
# FIX: parse data, don't execute it — ast.literal_eval for literals, or a
#   sandboxed expression evaluator; see the SAFE reference handler below.
# =============================================================================
@bp.post("/eval")
def eval_formula():
    expr = _field("expr")
    # VULNERABLE: eval() executes attacker-supplied text as Python code -> RCE.
    result = eval(expr)
    return jsonify({"expr": expr, "result": repr(result)})


# =============================================================================
# SAFE REFERENCE — data-only parsing, no code execution.
# json.loads / yaml.safe_load construct ONLY primitive types (dict/list/str/
# int/float/bool/None); the !!python/object and __reduce__ machinery is never
# reachable, so a hostile payload can at worst produce a benign parse error.
# Shown so the vulnerable/safe pair can be diffed by SAST tools and learners.
# =============================================================================
@bp.post("/safe")
def safe_load():
    data = _field("data")
    try:
        # SAFE: json.loads yields plain data, never executes anything.
        parsed = json.loads(data)
    except (ValueError, TypeError) as exc:
        # SAFE: yaml.safe_load refuses python/object tags -> ConstructorError, not RCE.
        try:
            parsed = yaml.safe_load(data)
        except yaml.YAMLError as yexc:
            return jsonify({"error": f"{exc}; {yexc}"}), 400
    return jsonify({"parsed_type": type(parsed).__name__, "parsed": parsed})

File: javascript/routes/deserialization.js
/**
 * Insecure Deserialization / Software & Data Integrity Failures —
 * OWASP 2021 A08 Software and Data Integrity Failures |
 * OWASP 2025 A08 Software and Data Integrity Failures.
 *
 * Same annotation format as the reference SQL-injection module: each
 * permutation carries a standard header describing the flaw, its CWE, an
 * exploitation-likelihood rating, a prevalence estimate, an example payload,
 * and the corresponding safe pattern. The dangerous sink on each handler is
 * flagged inline with `VULNERABLE:`.
 *
 * Sinks demonstrated (all genuine, all exploitable when the app runs):
 *   P1  node-serialize .unserialize()   -> RCE via _$$ND_FUNC$$_ IIFE   (CWE-502)
 *   P2  eval('(' + data + ')')          -> arbitrary code execution     (CWE-95)
 *   P3  hand-rolled recursive merge     -> prototype pollution __proto__ (CWE-1321)
 */
'use strict';

const express = require('express');
const serialize = require('node-serialize');

const router = express.Router();

// A textbook recursive merge — the exact shape found in countless "deepMerge",
// "extend", and config-loader helpers. It walks the source object and copies
// keys into the target; nested objects recurse. It performs NO key filtering,
// so an attacker-supplied `__proto__` key mutates Object.prototype. Used by
// PERMUTATION 3 below.
function unsafeMerge(target, source) {
  for (const key in source) {
    const value = source[key];
    if (value !== null && typeof value === 'object') {
      if (target[key] === null || typeof target[key] !== 'object') target[key] = {};
      unsafeMerge(target[key], value); // recurses into target['__proto__'] === Object.prototype
    } else {
      target[key] = value;
    }
  }
  return target;
}

// ===========================================================================
// PERMUTATION 1 — Native object deserialization of attacker data (node-serialize)
// OWASP 2021 A08 Software and Data Integrity Failures  ->  OWASP 2025 A08 Software and Data Integrity Failures
// CWE-502: Deserialization of Untrusted Data
// EXPLOITATION LIKELIHOOD: HIGH — unserialize() eval()s any function carried in
//   the `_$$ND_FUNC$$_` marker; a trailing `()` makes it an IIFE that fires at
//   parse time for reliable, deterministic pre-auth RCE. Needs the known
//   node-serialize gadget/marker, so slightly less turnkey than plain eval.
// PREVALENCE TODAY: greenfield RARE (nobody adds node-serialize to a new app;
//   JSON.parse is the default and the library is abandoned + npm-audit/SAST
//   flagged) | legacy/10yr tech-debt LOW (niche even historically, but survives
//   in old code that needed to serialize functions/live sessions).
// EXAMPLE:
//   curl -s localhost:3000/deserialization/native -H 'Content-Type: application/json' \
//     -d '{"payload":"{\"x\":\"_$$ND_FUNC$$_function(){return require('child_process').execSync('id').toString()}()\"}"}'
//   -> response echoes the `id` command output.
// FIX: never deserialize untrusted input into live objects; use JSON.parse
//   (data-only) and validate against a schema. See /deserialization/safe.
// ===========================================================================
router.post('/native', (req, res) => {
  const payload = (req.body && req.body.payload) || '';
  try {
    // VULNERABLE: unserialize() eval()s embedded _$$ND_FUNC$$_ functions.
    const obj = serialize.unserialize(payload);
    const preview = {};
    for (const k of Object.keys(obj || {})) {
      preview[k] = typeof obj[k] === 'function' ? '[function]' : obj[k];
    }
    res.json({ payload, result: preview });
  } catch (err) {
    res.status(500).json({ payload, error: err.message });
  }
});

// ===========================================================================
// PERMUTATION 2 — Unsafe "config" parsing via eval() (CWE-95 code injection)
// OWASP 2021 A08 Software and Data Integrity Failures  ->  OWASP 2025 A08 Software and Data Integrity Failures
// CWE-95: Improper Neutralization of Directives in Dynamically Evaluated Code ('Eval Injection')
// EXPLOITATION LIKELIHOOD: CRITICAL — wrapping user text in eval('(' + data + ')')
//   executes ANY JavaScript with zero gadget needed; a one-liner yields pre-auth
//   RCE, fully deterministic. The most turnkey sink in this file.
// PREVALENCE TODAY: greenfield LOW (eval is the canonical "never do this";
//   eslint no-eval and every SAST rule flag it; JSON.parse is idiomatic) |
//   legacy/10yr tech-debt MEDIUM (eval('('+resp+')') was the pre-JSON.parse way
//   to parse AJAX/JSONP responses and loosely-typed config, still in old code).
// EXAMPLE:
//   curl -s localhost:3000/deserialization/eval -H 'Content-Type: application/json' \
//     -d '{"data":"{ maxItems: 6*7 }"}'                       # -> config {maxItems:42}
//   curl -s localhost:3000/deserialization/eval -H 'Content-Type: application/json' \
//     -d '{"data":"require('child_process').execSync('id').toString()"}'   # -> RCE
// FIX: parse configuration with JSON.parse (data only). See /deserialization/safe.
// ===========================================================================
router.post('/eval', (req, res) => {
  const data = (req.body && req.body.data) || '';
  try {
    // VULNERABLE: attacker-controlled string evaluated as live JavaScript.
    const config = eval('(' + data + ')');
    res.json({ data, evaluated: '(' + data + ')', config });
  } catch (err) {
    res.status(500).json({ data, error: err.message });
  }
});

// ===========================================================================
// PERMUTATION 3 — Prototype pollution via recursive merge of user JSON
// OWASP 2021 A08 Software and Data Integrity Failures  ->  OWASP 2025 A08 Software and Data Integrity Failures
// CWE-1321: Improperly Controlled Modification of Object Prototype Attributes ('Prototype Pollution')
// EXPLOITATION LIKELIHOOD: HIGH — polluting Object.prototype via a `__proto__`
//   key is trivial and deterministic; every object app-wide inherits the value.
//   Direct impact ranges from DoS / auth-bypass (e.g. injected isAdmin) to RCE
//   when a downstream gadget (template engine, child_process options) reads it —
//   the escalation is what needs a gadget, the pollution itself does not.
// PREVALENCE TODAY: greenfield MEDIUM (hand-rolled deepMerge / config loaders /
//   query-string parsers keep reintroducing it; a very active CVE class since
//   2018 even though many libs are now hardened) | legacy/10yr tech-debt HIGH
//   (pre-2018 lodash/jQuery.extend and unguarded custom merges still shipping).
// EXAMPLE:
//   curl -s localhost:3000/deserialization/merge -H 'Content-Type: application/json' \
//     -d '{"json":"{\"__proto__\":{\"isAdmin\":true}}"}'
//   -> response.probe.isAdmin === true, read off a brand-new unrelated object.
// FIX: reject/strip __proto__|constructor|prototype keys, build with
//   Object.create(null) or Map, or use a hardened merge. See /deserialization/safe.
// ===========================================================================
router.post('/merge', (req, res) => {
  const raw = (req.body && req.body.json) || '{}';
  let source;
  try {
    source = JSON.parse(raw); // JSON.parse keeps __proto__ as an inert OWN key...
  } catch (err) {
    return res.status(400).json({ input: raw, error: err.message });
  }
  const before = Object.getOwnPropertyNames(Object.prototype);
  const target = {};
  // VULNERABLE: recursive merge honors a `__proto__` key and writes through to
  // Object.prototype, polluting every object in the process.
  unsafeMerge(target, source);
  const after = Object.getOwnPropertyNames(Object.prototype);
  const pollutedKeys = after.filter((k) => !before.includes(k));

  // Read the pollution back off a freshly-created, completely unrelated object:
  const probe = {};
  const leakedOnFreshObject = {};
  for (const k of pollutedKeys) leakedOnFreshObject[k] = probe[k];

  res.json({
    input: raw,
    merged: target,
    pollutedKeys,
    leakedOnFreshObject,
    probe: { isAdmin: probe.isAdmin, polluted: probe.polluted },
  });
});

// ===========================================================================
// SAFE REFERENCE — data-only parsing with a prototype-pollution guard. Shown so
// the vulnerable/safe pair can be diffed by SAST tooling and learners. JSON.parse
// yields inert data (no functions execute, no code runs); the merge guard drops
// the dangerous keys before any recursion.
// ===========================================================================
router.post('/safe', (req, res) => {
  const raw = (req.body && req.body.data) || '{}';
  let parsed;
  try {
    parsed = JSON.parse(raw); // SAFE: pure data, no code/function evaluation.
  } catch (err) {
    return res.status(400).json({ input: raw, error: err.message });
  }
  // SAFE merge: allow-list nothing dangerous — refuse the pollution vectors.
  const FORBIDDEN = new Set(['__proto__', 'constructor', 'prototype']);
  const target = Object.create(null); // no prototype to pollute
  for (const key of Object.keys(parsed)) {
    if (FORBIDDEN.has(key)) continue;
    target[key] = parsed[key];
  }
  res.json({ input: raw, parsed, safeMerged: { ...target } });
});

module.exports = { base: '/deserialization', router };
